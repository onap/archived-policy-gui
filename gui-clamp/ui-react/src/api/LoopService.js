/*-
 * ============LICENSE_START=======================================================
 * ONAP CLAMP
 * ================================================================================
 * Copyright (C) 2019 AT&T Intellectual Property. All rights reserved.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ============LICENSE_END============================================
 * ===================================================================
 *
 */

export default class LoopService {
	static getLoopNames() {
		return fetch(window.location.pathname + 'restservices/clds/v2/loop/getAllNames', { method: 'GET', credentials: 'same-origin' })
			.then(function (response) {
				console.debug("GetLoopNames response received: ", response.status);
				if (response.ok) {
					return response.json();
				} else {
					console.error("GetLoopNames query failed");
					return {};
				}
			})
			.catch(function (error) {
				console.error("GetLoopNames error received", error);
				return {};
			});
	}

	static createLoop(loopName, templateName) {
		return fetch(window.location.pathname + 'restservices/clds/v2/loop/create/' + loopName + '?templateName=' + templateName, {
			method: 'POST',
			headers: {
				"Content-Type": "application/json"
			},
			credentials: 'same-origin'
		})
			.then(function (response) {
				console.debug("CreateLoop response received: ", response.status);
				return response.json();
			})
			.catch(function (error) {
				console.error("CreateLoop error received", error);
				return "";
			});
	}

	static getLoop(loopName) {
		return fetch(window.location.pathname + 'restservices/clds/v2/loop/' + loopName, {
			method: 'GET',
			headers: {
				"Content-Type": "application/json"
			},
			credentials: 'same-origin'
		})
			.then(function (response) {
				console.debug("GetLoop response received: ", response.status);
				if (response.ok) {
					return response.json();
				} else {
					console.error("GetLoop query failed");
					return {};
				}
			})
			.catch(function (error) {
				console.error("GetLoop error received", error);
				return {};
			});
	}

	static setMicroServiceProperties(loopName, jsonData) {
		return fetch(window.location.pathname + 'restservices/clds/v2/loop/updateMicroservicePolicy/' + loopName, {
			method: 'POST',
			credentials: 'same-origin',
			headers: {
				"Content-Type": "application/json"
			},
			body: JSON.stringify(jsonData)
		})
			.then(function (response) {
				console.debug("updateMicroservicePolicy response received: ", response.status);
				if (response.ok) {
					return response.text();
				} else {
					console.error("updateMicroservicePolicy query failed");
					return "";
				}
			})
			.catch(function (error) {
				console.error("updateMicroservicePolicy error received", error);
				return "";
			});
	}
	
	static setOperationalPolicyProperties(loopName, jsonData) {
		return fetch(window.location.pathname + 'restservices/clds/v2/loop/updateOperationalPolicies/' + loopName, {
			method: 'POST',
			credentials: 'same-origin',
			headers: {
				"Content-Type": "application/json"
			},
			body: JSON.stringify(jsonData)
		})
			.then(function (response) {
				console.debug("updateOperationalPolicies response received: ", response.status);
				if (response.ok) {
					return response.text();
				} else {
					console.error("updateOperationalPolicies query failed");
					return "";
				}
			})
			.catch(function (error) {
				console.error("updateOperationalPolicies error received", error);
				return "";
			});
	}

	static updateGlobalProperties(loopName, jsonData) {
		return fetch(window.location.pathname + 'restservices/clds/v2/loop/updateGlobalProperties/' + loopName, {
			method: 'POST',
			credentials: 'same-origin',
			headers: {
				"Content-Type": "application/json"
			},
			body: JSON.stringify(jsonData)
		})
			.then(function (response) {
				console.debug("updateGlobalProperties response received: ", response.status);
				if (response.ok) {
					return response.text();
				} else {
					console.error("updateGlobalProperties query failed");
					return "";
				}
			})
			.catch(function (error) {
				console.error("updateGlobalProperties error received", error);
				return "";
			});
	}

	static refreshOperationalPolicyJson(loopName,operationalPolicyName) {
		return fetch(window.location.pathname + 'restservices/clds/v2/loop/refreshOperationalPolicyJsonSchema/' + loopName + '/' + operationalPolicyName, {
			method: 'PUT',
			headers: {
				"Content-Type": "application/json"
			},
			credentials: 'same-origin'
		})
			.then(function (response) {
				console.debug("Refresh Operational Policy Json Schema response received: ", response.status);
				if (response.ok) {
					return response.json();
				} else {
					console.error("Refresh Operational Policy Json Schema query failed");
					return {};
				}
			})
			.catch(function (error) {
				console.error("Refresh Operational Policy Json Schema error received", error);
				return {};
			});
	}

		static refreshMicroServicePolicyJson(loopName,microServicePolicyName) {
    		return fetch(window.location.pathname + 'restservices/clds/v2/loop/refreshMicroServicePolicyJsonSchema/' + loopName + '/' + microServicePolicyName, {
    			method: 'PUT',
    			headers: {
    				"Content-Type": "application/json"
    			},
    			credentials: 'same-origin'
    		})
    			.then(function (response) {
    				console.debug("Refresh Operational Policy Json Schema response received: ", response.status);
    				if (response.ok) {
    					return response.json();
    				} else {
    					console.error("Refresh Operational Policy Json Schema query failed");
    					return {};
    				}
    			})
    			.catch(function (error) {
    				console.error("Refresh Operational Policy Json Schema error received", error);
    				return {};
    			});
    }

	static addOperationalPolicyType(loopName, policyType, policyVersion) {
		return fetch(window.location.pathname + 'restservices/clds/v2/loop/addOperationaPolicy/' + loopName + '/policyModel/' + policyType +'/' + policyVersion , {
			method: 'PUT',
			headers: {
				"Content-Type": "application/json"
			},
			credentials: 'same-origin'
		})
				.then(function (response) {
				console.debug("Add Operational Policy response received: ", response.status);
				if (response.ok) {
					return response.json();
				} else {
					return response.text();
				}
			})
			.catch(function (error) {
				console.error("Add Operational Policy query failed");
				throw new Error(error);
			})
	}

	static removeOperationalPolicyType(loopName, policyType, policyVersion, policyName) {
		return fetch(window.location.pathname + 'restservices/clds/v2/loop/removeOperationaPolicy/' + loopName + '/policyModel/' + policyType +'/' + policyVersion + '/' + policyName , {
			method: 'PUT',
			headers: {
				"Content-Type": "application/json"
			},
			credentials: 'same-origin'
		})
				.then(function (response) {
					console.debug("Remove Operational Policy response received: ", response.status);
				if (response.ok) {
					return response.json();
				} else {
					console.error("Remove Operational Policy query failed");
					return {};
				}
			})
			.catch(function (error) {
				console.error("Remove Operational Policy error received", error);
				return {};
			});
	}
}
