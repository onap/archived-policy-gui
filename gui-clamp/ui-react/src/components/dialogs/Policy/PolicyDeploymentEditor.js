/*-
 * ============LICENSE_START=======================================================
 * ONAP POLICY-CLAMP
 * ================================================================================
 * Copyright (C) 2021 AT&T Intellectual Property. All rights
 *                             reserved.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ============LICENSE_END============================================
 * ===================================================================
 *
 */

import React, { forwardRef } from 'react';
import Modal from 'react-bootstrap/Modal';
import styled from 'styled-components';
import Button from 'react-bootstrap/Button';
import Alert from 'react-bootstrap/Alert';
import PolicyService from '../../../api/PolicyService';
import FormGroup from '@material-ui/core/FormGroup';
import Checkbox from '@material-ui/core/Checkbox';
import FormControlLabel from '@material-ui/core/FormControlLabel';

const DivWhiteSpaceStyled = styled.div`
  white-space: pre;
`

const PanelDiv = styled.div`
  text-align: justify;
  font-size: ${ props => props.theme.policyEditorFontSize };
  background-color: ${ props => props.theme.loopViewerBackgroundColor };
`

export default class PolicyDeploymentEditor extends React.Component {

  state = {
    policyData: this.props.policyData,
    showSuccessAlert: false,
    showFailAlert: false,
    checkboxesState: this.createPdpStructure(this.props.policyData),
    checkboxesInitialState: this.createPdpStructure(this.props.policyData),
  };

  constructor(props, context) {
    super(props, context);
    this.handleClose = this.handleClose.bind(this);
    this.handleUpdatePdpDeployment = this.handleUpdatePdpDeployment.bind(this);
    this.disableAlert = this.disableAlert.bind(this);
    this.renderPdpDeploymentCheckboxes = this.renderPdpDeploymentCheckboxes.bind(this);
    this.createPdpStructure = this.createPdpStructure.bind(this);
    this.handleCheckboxChange = this.handleCheckboxChange.bind(this);
    this.createPdpGroupOperations = this.createPdpGroupOperations.bind(this);
  }

  handleClose() {
    this.setState({ show: false });

  }

  disableAlert() {
    this.setState({ showSuccessAlert: false, showFailAlert: false });
  }

  createPdpGroupOperations(initialStates, newStates) {
    let commandsArray = [];
    initialStates.forEach(initElem => {
      let newStateFound = newStates.find(newElement => newElement.name === initElem.name);
      if (initElem.value !== newStateFound.value) {
        let newPdpGroupsArray = newStateFound.name.split("/");
        let operation = "POST/";
        if (initElem.value) {
          operation = "DELETE/";
        }
        commandsArray.push(operation + newPdpGroupsArray[0] + "/" + newPdpGroupsArray[1] + "/"
          + this.state.policyData.name + "/" + this.state.policyData.version);
      }
    });
    return commandsArray.length > 0 ? { "PdpActions": commandsArray } : undefined;
  }

  handleUpdatePdpDeployment() {
    let operationsList = this.createPdpGroupOperations(this.state.checkboxesInitialState,
      this.state.checkboxesState);
    if (typeof (operationsList) !== "undefined") {
      PolicyService.updatePdpDeployment(operationsList).then(respPdpDeploymentUpdate => {
        if (typeof (respPdpDeploymentUpdate) === "undefined") {
          //it indicates a failure
          this.setState({
            showFailAlert: true,
            showMessage: 'Pdp Deployment update Failure'
          });
        } else {
          this.setState({
            showSuccessAlert: true,
            showMessage: 'Pdp Deployment Update successful'
          });
          this.props.policiesTableUpdateFunction();
        }
      })
    } else {
      this.setState({
        showSuccessAlert: true,
        showMessage: 'Pdp Deployment: Nothing to change'
      });
    }
  }

  createPdpStructure(policyData) {
    // Create map with data for all group/subgroup where the policy is deployed
    let infoPdpMap = new Map();
    if (typeof policyData.pdpGroupInfo !== "undefined") {
      policyData["pdpGroupInfo"].forEach(pdpGroupElem => {
        let pdpGroupName = Object.keys(pdpGroupElem)[0];
        pdpGroupElem[pdpGroupName]["pdpSubgroups"].forEach(pdpSubGroupElem => {
          infoPdpMap.set(pdpGroupName + "/" + pdpSubGroupElem["pdpType"], true);
        });
      });
    }
    // Create the possible values for pdpgroup/subgroup and tick the ones where policy is deployed
    let pdpStates = [];
    if (typeof policyData.supportedPdpGroups !== "undefined") {
      for (const pdpGroup of policyData["supportedPdpGroups"]) {
        let pdpGroupName = Object.keys(pdpGroup)[0];
        for (const pdpSubGroup of Object.values(pdpGroup)[0]) {
          let fullName = pdpGroupName + "/" + pdpSubGroup;
          pdpStates.push({
            name: fullName,
            value: infoPdpMap.get(fullName) !== undefined
          });
        }
      }
    }
    return pdpStates;
  }

  handleCheckboxChange(event) {
    const checkboxesArray = this.state.checkboxesState;
    checkboxesArray.find(element => element.name === event.target.name).value = event.target.checked;
    this.setState({ checkboxesState: checkboxesArray });
  }

  renderPdpDeploymentCheckboxes() {
    return this.state.checkboxesState.map(item => {
      return <FormControlLabel control={ <Checkbox checked={ item.value } onChange={ this.handleCheckboxChange }
                                                   name={ item.name }/> } label={ item.name }/>;
    });
  }

  render() {
    return (
      <PanelDiv>
        <Alert variant="success" show={ this.state.showSuccessAlert } onClose={ this.disableAlert } dismissible>
          <DivWhiteSpaceStyled>
            { this.state.showMessage }
          </DivWhiteSpaceStyled>
        </Alert>
        <Alert variant="danger" show={ this.state.showFailAlert } onClose={ this.disableAlert } dismissible>
          <DivWhiteSpaceStyled>
            { this.state.showMessage }
          </DivWhiteSpaceStyled>
        </Alert>
        <Button variant="secondary" title="Update the policy to the specified PDP Groups/Subgroups"
                onClick={ this.handleUpdatePdpDeployment }>Update PDP</Button>
        <FormGroup>{ this.renderPdpDeploymentCheckboxes() }</FormGroup>
      </PanelDiv>
    );
  }
}
