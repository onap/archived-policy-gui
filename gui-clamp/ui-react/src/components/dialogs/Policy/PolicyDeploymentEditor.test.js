/*
 * ============LICENSE_START=======================================================
 *  Copyright (C) 2022 Nordix Foundation.
 *  ================================================================================
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *  SPDX-License-Identifier: Apache-2.0
 *  ============LICENSE_END=========================================================
 */

import {shallow} from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";
import PolicyDeploymentEditor from "./PolicyDeploymentEditor";
import fs from "fs";
import LoopService from "../../../api/LoopService";

describe('Verify PolicyDeploymentEditor', () => {
    const toscaPolicyDeploymentEditor = fs.readFileSync('src/components/dialogs/Policy/toscaPolicyDeploymentEditor.test.json', {
        encoding: 'utf8',
        flag: 'r'
    });

    const toscaPolicyDeploymentEditorArray = JSON.parse(toscaPolicyDeploymentEditor);

    const initialStateTrue = [
        {
            "name": "monitoring/xacml",
            "value": true
        }
    ];

    const initialStateFalse = [
        {
            "name": "monitoring/xacml",
            "value": false
        }
    ];

    const logSpy = jest.spyOn(console, 'log');

    it("renders correctly", () => {
        const component = shallow(<PolicyDeploymentEditor policyData={toscaPolicyDeploymentEditorArray}/>);
        expect(toJson(component)).toMatchSnapshot();
    });

    it('Test createPdpGroupOperations', () => {
        const component = shallow(<PolicyDeploymentEditor policyData={toscaPolicyDeploymentEditorArray} />);

        const instance = component.instance();

        instance.createPdpGroupOperations(initialStateTrue, initialStateTrue);

        component.update();

        expect(logSpy).toHaveBeenCalledWith('createPdpGroupOperations called');
    });

    it('Test handleUpdatePdpDeployment', () => {
        LoopService.updatePdpDeployment = jest.fn().mockImplementation(() => {
            return Promise.resolve(undefined);
        });

        const component = shallow(<PolicyDeploymentEditor policyData={toscaPolicyDeploymentEditorArray} />);
        component.setState({checkboxesInitialState: initialStateTrue});
        component.setState({checkboxesState: initialStateFalse});

        const instance = component.instance();
        instance.handleUpdatePdpDeployment();

        expect(component.state('showFailAlert')).toEqual(false);
        expect(component.state('showMessage')).toEqual(undefined);

        component.update();

        component.setState({showFailAlert: true});
        component.setState({showMessage: 'Pdp Deployment update Failure'});

        expect(logSpy).toHaveBeenCalledWith('handleUpdatePdpDeployment called');
        expect(component.state('showFailAlert')).toEqual(true);
        expect(component.state('showMessage')).toEqual('Pdp Deployment update Failure');
    });

});