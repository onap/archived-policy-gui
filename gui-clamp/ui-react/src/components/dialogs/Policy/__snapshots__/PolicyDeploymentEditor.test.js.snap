// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify PolicyDeploymentEditor renders correctly 1`] = `
<styled.div>
  <Alert
    closeLabel="Close alert"
    dismissible={true}
    onClose={[Function]}
    show={false}
    transition={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "appear": false,
          "in": false,
          "mountOnEnter": false,
          "timeout": 300,
          "unmountOnExit": false,
        },
        "render": [Function],
      }
    }
    variant="success"
  >
    <styled.div />
  </Alert>
  <Alert
    closeLabel="Close alert"
    dismissible={true}
    onClose={[Function]}
    show={false}
    transition={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "appear": false,
          "in": false,
          "mountOnEnter": false,
          "timeout": 300,
          "unmountOnExit": false,
        },
        "render": [Function],
      }
    }
    variant="danger"
  >
    <styled.div />
  </Alert>
  <Button
    active={false}
    disabled={false}
    onClick={[Function]}
    title="Update the policy to the specified PDP Groups/Subgroups"
    variant="secondary"
  >
    Update PDP
  </Button>
  <WithStyles(ForwardRef(FormGroup))>
    <WithStyles(ForwardRef(FormControlLabel))
      control={
        <WithStyles(ForwardRef(Checkbox))
          checked={true}
          name="monitoring/xacml"
          onChange={[Function]}
        />
      }
      label="monitoring/xacml"
    />
  </WithStyles(ForwardRef(FormGroup))>
</styled.div>
`;
